File: ClientJava/src/main/java/Proxy/ProtoUtils.java
package Proxy;

import Domain.AgeCategory;
import Domain.Participant;
import Domain.Trial;
import Domain.User;
import Generated.Protobuf.AgeCategoryDTO;
import Generated.Protobuf.ParticipantDTO;
import Generated.Protobuf.TrialDTO;
import Generated.Protobuf.UserDTO;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by andrei on 2017-04-29.
 */
public class ProtoUtils {
    public static User userFromDto(UserDTO userDto) {
        return new User(userDto.getUserName(), userDto.getPassword());
    }

    public static UserDTO userToDto(User user) {
        return UserDTO.newBuilder().setUserName(user.getUserName()).setPassword(user.getPassword()).build();
    }

    public static AgeCategory ageCategoryFromDto(AgeCategoryDTO ageCategoryDTO) {
        return new AgeCategory(ageCategoryDTO.getName(), ageCategoryDTO.getMinAge(), ageCategoryDTO.getMaxAge());
    }

    public static AgeCategoryDTO ageCategoryToDto(AgeCategory ageCategory) {
        return AgeCategoryDTO.newBuilder().setName(ageCategory.getName()).setMaxAge(ageCategory.getMaxAge()).setMinAge(ageCategory.getMinAge()).build();
    }


    public static Participant participantFromDto(ParticipantDTO participantDTO) {
        return new Participant(participantDTO.getName(), participantDTO.getAge(), ageCategoryFromDto(participantDTO.getAgeCategory()));
    }

    public static ParticipantDTO participantToDto(Participant participant) {
        return ParticipantDTO.newBuilder().setName(participant.getName()).setAge(participant.getAge()).setAgeCategory(ageCategoryToDto(participant.getAgeCategory())).build();
    }

    public static Trial trialFromDto(TrialDTO trialDTO) {
        return new Trial(trialDTO.getName(), trialDTO.getDifficulty());
    }

    public static TrialDTO trialToDto(Trial trial) {
        return TrialDTO.newBuilder().setName(trial.getName()).setDifficulty(trial.getDifficulty()).build();
    }

    public static List<AgeCategoryDTO> ageCategoryListToDto(List<AgeCategory> ageCategoryList) {
        List<AgeCategoryDTO> all = new ArrayList<>();
        for (AgeCategory a : ageCategoryList) all.add(ageCategoryToDto(a));
        return all;
    }

    public static List<AgeCategory> ageCategoryListFromDto(List<AgeCategoryDTO> ageCategoryDtoList) {
        List<AgeCategory> all = new ArrayList<>();
        for (AgeCategoryDTO a : ageCategoryDtoList) all.add(ageCategoryFromDto(a));
        return all;
    }

    public static List<TrialDTO> trialsListToDto(List<Trial> trialsList) {
        List<TrialDTO> all = new ArrayList<>();
        for (Trial t : trialsList) all.add(trialToDto(t));
        return all;
    }

    public static List<Trial> trialsFromDto(List<TrialDTO> trialDTOList) {
        List<Trial> all = new ArrayList<>();
        for (TrialDTO t : trialDTOList) all.add(trialFromDto(t));
        return all;
    }

    public static List<ParticipantDTO> participantsListToDto(List<Participant> participantList) {
        List<ParticipantDTO> all = new ArrayList<>();
        for (Participant t : participantList) all.add(participantToDto(t));
        return all;
    }

    public static List<Participant> participantsListFromDto(List<ParticipantDTO> participantDtoList) {
        List<Participant> all = new ArrayList<>();
        for (ParticipantDTO t : participantDtoList) all.add(participantFromDto(t));
        return all;
    }
}

File: ServerCSharp/ServerCSharp/Protobuf/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Generated.Protobuf;
using ServerCSharp.Domain;

namespace ServerCSharp.Protobuf
{
   public static class ProtoUtils
   {
      public static AgeCategoryDTO AgeCategoryToDto(AgeCategory ageCategory)
      {
         return new AgeCategoryDTO {MaxAge = ageCategory.MaxAge, MinAge = ageCategory.MinAge, Name = ageCategory.Name};
      }

      public static AgeCategory AgeCategoryFromDto(AgeCategoryDTO ageCategoryDto)
      {
         return new AgeCategory(ageCategoryDto.Name, ageCategoryDto.MinAge, ageCategoryDto.MaxAge);
      }

      public static UserDTO UserToDto(User user)
      {
         return new UserDTO {Password = user.Password, UserName = user.UserName};
      }

      public static User UserFromDto(UserDTO userDto)
      {
         return new User(userDto.UserName, userDto.Password);
      }

      public static TrialDTO TrialToDto(Trial trial)
      {
         return new TrialDTO {Name = trial.Name, Difficulty = trial.Difficulty};
      }

      public static Trial TrialFromDto(TrialDTO trialDto)
      {
         return new Trial(trialDto.Name, trialDto.Difficulty);
      }

      public static ParticipantDTO ParticipantToDto(Participant participant)
      {
         return new ParticipantDTO
         {
            Age = participant.Age,
            AgeCategory = AgeCategoryToDto(participant.AgeCategory),
            Name = participant.Name
         };
      }

      public static Participant ParticipantFromDto(ParticipantDTO participantDto)
      {
         return new Participant(participantDto.Name, participantDto.Age, AgeCategoryFromDto(participantDto.AgeCategory));
      }

      public static List<AgeCategoryDTO> AgeCategoryListToDto(List<AgeCategory> ageCategoriesList)
      {
         List<AgeCategoryDTO> resultedList = new List<AgeCategoryDTO>();
         ageCategoriesList.ForEach(o => { resultedList.Add(AgeCategoryToDto(o)); });
         return resultedList;
      }

      public static List<AgeCategory> AgeCategoryListFromDto(List<AgeCategoryDTO> ageCategoriesDtoList)
      {
         List<AgeCategory> resultedList = new List<AgeCategory>();
         ageCategoriesDtoList.ForEach(o => { resultedList.Add(AgeCategoryFromDto(o)); });
         return resultedList;
      }

      public static List<TrialDTO> TrialsListToDto(List<Trial> trialsList)
      {
         List<TrialDTO> resultedList = new List<TrialDTO>();
         trialsList.ForEach(o => { resultedList.Add(TrialToDto(o)); });
         return resultedList;
      }

      public static List<Trial> TrialsListFromDto(List<TrialDTO> trialsDtoList)
      {
         List<Trial> resultedList = new List<Trial>();
         trialsDtoList.ForEach(o => { resultedList.Add(TrialFromDto(o)); });
         return resultedList;
      }

      public static List<ParticipantDTO> ParticipantsListToDto(List<Participant> participantsList)
      {
         List<ParticipantDTO> resultedList = new List<ParticipantDTO>();
         participantsList.ForEach(o => { resultedList.Add(ParticipantToDto(o)); });
         return resultedList;
      }
      public static List<Participant> ParticipantsListFromDto(List<ParticipantDTO> participantsDtoList)
      {
         List<Participant> resultedList = new List<Participant>();
         participantsDtoList.ForEach(o => { resultedList.Add(ParticipantFromDto(o)); });
         return resultedList;
      }
   }
}